src/paas_charm/charm_state.py

```diff
@@ -241,35 +225,14 @@ def build(  # pylint: disable=too-many-arguments
 
         Return:
             The IntegrationsState instance created.
-
-        Raises:
-            CharmConfigInvalidError: If some parameter in invalid.
         """
-        if s3_connection_info:
-            try:
-                # s3_connection_info is not really a Dict[str, str] as stated in
-                # charms.data_platform_libs.v0.s3. It is really a
-                # Dict[str, str | list[str]].
-                # Ignoring as mypy does not work correctly with that information.
-                s3_parameters = S3Parameters(**s3_connection_info)  # type: ignore[arg-type]
-            except ValidationError as exc:
-                error_message = build_validation_error_message(exc)
-                raise CharmConfigInvalidError(
-                    f"Invalid S3 configuration: {error_message}"
-                ) from exc
-        else:
-            s3_parameters = None
-
-        if saml_relation_data is not None:
-            try:
-                saml_parameters = SamlParameters(**saml_relation_data)
-            except ValidationError as exc:
-                error_message = build_validation_error_message(exc)
-                raise CharmConfigInvalidError(
-                    f"Invalid Saml configuration: {error_message}"
-                ) from exc
-        else:
-            saml_parameters = None
+        s3_parameters = typing.cast(
+            S3Parameters | None, generate_relation_parameters(s3_connection_info, S3Parameters)
+        )
+        saml_parameters = typing.cast(
+            SamlParameters | None,
+            generate_relation_parameters(saml_relation_data, SamlParameters, True),
+        )
```

Comment:
I think I don't think the type casting is necessary here since a `TypeVar` is being used to define the argument type?